### PR TITLE
Implement and extend `mul(sparse, sparse)` to work with 0-dim arguments on either side.

### PR DESCRIPTION
**Pull request by nikitaved**
Extends https://github.com/pytorch/pytorch/pull/81556 by bringing some missing functionality implemented in master.
Also, improves on master to allow arbitrary 0-dim coalesced or not arguments to be on either side of the operation.
Master, for example, would fail on 0-dim non-coalesced inputs.

CC @datumbox, @osalpekar .

### CODE REVIEW

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp
Comment (nikitaved) [inline aten/src/ATen/native/sparse/SparseTensorMath.cpp:942]: We won't need these checks with a better version of `mul(sparse, sparse)`.

**Comment by nikitaved** [comment]
Dealing with 0-dim tensors, especially if they are sparse, is a real pain...

File: aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu
Comment (amjames) [inline aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:486]: The nested ternary with compound conditions is a bit much to digest, but I tried and is probably the clearest way to write this. I suggest a comment describing these two assignments so someone does not have to manually pick it apart if they are reading. Same applies to the block in the .cpp

**Review by amjames — APPROVED** [review]
One suggestion, see inline comments. Otherwise LGTM.

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp
Comment (bhosmer) [inline aten/src/ATen/native/sparse/SparseTensorMath.cpp:961]: re lines 946-961 - I totally get the desire to avoid big branchy if/elseif nests, but the logic here seems kind of complicated and I think there's some avoidable redundancy. 

WDYT of something like this? Assuming I've read the original correctly, it does the same thing and avoids the double `!dim()` testing and use of a fresh Tensor as a "none" value:
```
if (!src_.dim()) {
  vals = get_vals_from_scalar(src_.is_coalesced() ? src_ : src_.coalesce());
  return _mul_dense_sparse_out(vals, t_, r);
}
if (!t_.dim()) {
  vals = get_vals_from_scalar(t_.is_coalesced() ? t_ : t_.coalesce());
  return _mul_dense_sparse_out(vals, src_, r);
}
```

File: aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu
Comment (bhosmer) [inline aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:498]: Ditto comments from CPU version 😁

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp
Comment (bhosmer) [inline aten/src/ATen/native/sparse/SparseTensorMath.cpp:941]: when would this happen? (just point me to a test case if you've got one)

**Review by bhosmer — COMMENTED** [review]
Looks good, thanks Nik! Couple comments inline but only one immediate suggestion - happy to go over this when we meet tomorrow also.

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp
Comment (nikitaved) [reply aten/src/ATen/native/sparse/SparseTensorMath.cpp:801]: Right!

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp
Comment (nikitaved) [reply aten/src/ATen/native/sparse/SparseTensorMath.cpp:961]: Yeah, I like the suggestion, but I wanted to postpone the call to coalesce as much as possible, as it is expensive (sync plus synchronizations).

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp
Comment (nikitaved) [reply aten/src/ATen/native/sparse/SparseTensorMath.cpp:809]: Makes sense. I thought this approach is axiomatic as it is quite ubiquitous.

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp
Comment (amjames) [reply aten/src/ATen/native/sparse/SparseTensorMath.cpp:961]: Shouldn't coalesce short circuit to a no-op if `is_coalesced()`?

Comment (nikitaved) [reply aten/src/ATen/native/sparse/SparseTensorMath.cpp:961]: From the code suggested it could be that src is not coalesced and t is coalesced, and we end up calling coalesce when it could be avoided.

Comment (bhosmer) [reply aten/src/ATen/native/sparse/SparseTensorMath.cpp:961]: I think the simplification would call coalesce in the same places/at the same time as the original, no? Maybe I'm missing it - sorry to make you spell it out, but can you give me an example where the coalesce behavior would be different?

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp
Comment (nikitaved) [reply aten/src/ATen/native/sparse/SparseTensorMath.cpp:941]: Yes it happens in test_basic_ops. You can have a sparse tensor of shape () with values of dim 1 and shape [0].

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp
Comment (nikitaved) [reply aten/src/ATen/native/sparse/SparseTensorMath.cpp:961]: Imagine that src is non-coalesced, t is coalesced, and both of them are 0-dim.

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp
Comment (bhosmer) [reply aten/src/ATen/native/sparse/SparseTensorMath.cpp:809]: yeah same, I've read past it dozens of times without checking 😛 back when we did an overhead reduction sweep we realized these tiny functions with asserts tucked into them were defeating a ton of inlining we'd assumed was happening

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp
Comment (bhosmer) [reply aten/src/ATen/native/sparse/SparseTensorMath.cpp:961]: Ah got it, thanks! And yeah, agree it's probably worth using an optional rather than the Tensor {} sentinel.

**Comment by nikitaved** [comment]
Thank you, @amjames , @bhosmer , for your comments. I have refactored the code. Hope it is easier to understand now.

**Review by amjames — APPROVED** [review]
Big fan of the refactor! Awesome work.

**Review by bhosmer — APPROVED** [review]
Yeah agree with @amjames, this refactor and the excellent comments make it so much easier to follow the multiple constraints the code is solving for. Gold in terms of maintainability 😁 thanks a bunch for doing this!

**Comment by nikitaved** [comment]
Thank you :) Let's hope it solves the problem with torchvision.

@pytorchbot merge

**Comment by nikitaved** [comment]
@datumbox , would it be easier now to check whether the issue is gone? Or shall we wait until it goes into the nightly build?

**Comment by kit1980** [comment]
@nikitaved heads up we may need to revert this and https://github.com/pytorch/pytorch/pull/81556 because of internal torchvision breakage. Will have more clarity tomorrow.

**Comment by datumbox** [comment]
@kit1980 could you confirm that the 2 PRs have been reverted? I just want to be mindful that if we leave https://github.com/pytorch/pytorch/pull/81556 unreverted for too long it might become too difficult to change due to conflicts.

@nikitaved I don't have much context or visibility about the work on these PRs. Given these series of PRs have caused issues on downstream projects, it might be worth asking from the Meta employee that supports you to merge it manually on Phabricator rather than doing it with a bot. This will help them identify any breakages before hand easier.

**Comment by nikitaved** [comment]
Right, but I would like to know the source of the issue. It could be that some tests expect failures and they do not happen.

**Comment by datumbox** [comment]
@nikitaved Make sense. The person who is supporting you should be able to give you all these details along with a proposal on how to resolve.

I got pinged here because of the potential breakage downstream and I'm missing the context to be able to give the such support.

**Comment by kit1980** [comment]
@datumbox @nikitaved Not reverted yet, but will probably revert soon to avoid further github/internal divergence. Unfortunately not clear why the internal torchvision is failing.
If this is reverted, please re-submit the PRs but do not merge and ping me.

**Comment by nikitaved** [comment]
I have seen some logs, they show JIT failures with image transformations. Do these even use sparse tensors really?

**Comment by kit1980** [comment]
@nikitaved I had to revert this, reverting the previous one also.
If you can combine this and https://github.com/pytorch/pytorch/pull/81556 into one PR and re-submit, it would be great.
Please add me before GitHub merge so we can verify internally before merge.

**Pull request closed without merging**
